### PR TITLE
Reduce data races in the debugger

### PR DESCRIPTION
**Pull request by slavek-kucera**

### CODE REVIEW

File: parser_library/test/debugging/debug_lib_provider_test.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    const std::string aaa_content = " MNOTE 'AAA'";
+    const resource_location aaa_locaiton("AAA");
+    EXPECT_CALL(*mock_lib, get_file_content(Eq("AAA"))).WillOnce(Return(std::pair(aaa_locaiton, aaa_content)));
+
+    std::string input = " COPY AAA";
+    analyzer a(input, analyzer_options(&lib));
+    a.analyze();
+    a.collect_diags();
+
+    EXPECT_TRUE(matches_message_codes(a.diags(), { "MNOTE" }));
```
Comment (jirimosinger) [inline parser_library/test/debugging/debug_lib_provider_test.cpp:66]:
```suggestion
    const std::string aaa_content = " MNOTE 'AAA'";
    const resource_location aaa_location("AAA");
    EXPECT_CALL(*mock_lib, get_file_content(Eq("AAA"))).WillOnce(Return(std::pair(aaa_location, aaa_content)));

    std::string input = " COPY AAA";
    analyzer a(input, analyzer_options(&lib));
    a.analyze();
    a.collect_diags();

    EXPECT_TRUE(matches_message_codes(a.diags(), { "MNOTE" }));
```

File: parser_library/src/workspaces/library_local.h [hunk 2]
```diff
@@ -51,34 +54,57 @@ class library_local final : public library, public diagnosable_impl
...
+        auto exchange(state_t data) { return std::atomic_exchange(&m_data, std::move(data)); }
+    };
+#endif
+
     file_manager& m_file_manager;
 
     utils::resource::resource_location m_lib_loc;
-    std::unordered_map<std::string, utils::resource::resource_location, utils::hashers::string_hasher, std::equal_to<>>
-        m_files;
+    atomic_state_t m_files;
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply parser_library/src/workspaces/library_local.h:100]: The atomic_state_t should not exist at all, but both libc++ and libstdc++ still do not have `std::atomic<std::shared_ptr<T>>` support.
As for `state_t` I am open to suggestions.

[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/workspaces/library_local.h:100]: What about something from the list:
- `files_collection_t`
- `files_store_t`
- `library_information_t`
- ...

File: clients/vscode-hlasmplugin/src/test/suite/integration.test.ts [hunk 3]
```diff
@@ -38,12 +37,7 @@ suite('Integration Test Suite', () => {
     // change 'open' file to create diagnostic
     test('Diagnostic test', async () => {
         // register callback to check for the correctness of the diagnostic
-        const diagnostic_event = new Promise<[vscode.Uri, vscode.Diagnostic[]][]>((resolve, reject) => {
-            const listener = vscode.languages.onDidChangeDiagnostics((_) => {
-                listener.dispose();
-                resolve(vscode.languages.getDiagnostics());
-            });
-        });
+        const diagnostic_event = helper.waitForDiagnostics();
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply clients/vscode-hlasmplugin/src/test/suite/integration.test.ts:40]: nope... I need to start the waiting process (hook the event), trigger the action that will produce the diags and only then await it.

**Review by jirimosinger — APPROVED** [review]

**Review by jirimosinger — APPROVED** [review]

**Pull request merged**